### PR TITLE
Port 7 high-value problem variants (new vehicle_routing category + knapsack/network_flow/portfolio/assignment/facility_location)

### PR DESCRIPTION
**Pull request by flixpar**
## Summary

Ports and upgrades **seven high-value problem variants** identified in `docs/variant_branch_review.md` that had not yet been integrated into `main`. Adds one new category (`vehicle_routing`) and diversifies four previously single-variant categories (`knapsack`, `network_flow`, `portfolio`, `assignment`) plus a classic addition to `facility_location`. `list_problems()` grows from **49 → 56** registered `category/variant` pairs.

Each variant was reimplemented to the project quality bar (scale-tiered realistic data, careful `target_variables` scaling, provable feasibility) rather than copied — every reviewer-noted defect on the source branches (degenerate LP relaxations, no-op infeasibility handling) was fixed. Infeasibility is constructed to hold in the **LP relaxation**, since the framework relaxes integrality by default.

## What's added

| Variant | What it is | Notable upgrade vs. the source branch |
|---|---|---|
| **`vehicle_routing/cvrp`** 🆕 *new category* | Capacitated Vehicle Routing | Gavish–Graves single-commodity-flow subtour elimination → depot-anchored, non-degenerate relaxation |
| **`knapsack/multidimensional`** | Multi-constraint 0/1 knapsack | Structural min-items covering-floor infeasibility (replaces a no-op) |
| **`knapsack/bounded`** | Integer-multiplicity knapsack | Value floor above the exact bounded-fractional greedy LP optimum |
| **`network_flow/generalized_flow`** | Lossy flow with per-arc gains | Min-cost delivery (not degenerate source-outflow max) |
| **`portfolio/tracking_error`** | Index tracking (MAD tracking-error budget) | Benchmark-feasible by construction; `cvar` stays default |
| **`assignment/workload_balance`** | Minimax/makespan assignment | Capacity-pigeonhole infeasibility; distinct from `load_balancing` |
| **`facility_location/p_median`** | Classic p-median | Tight disaggregated linking + count-capacity pigeonhole |

## Verification

- **~2,300 independent HiGHS solves** across targets 6–2000 and many seeds: every `feasible` → OPTIMAL, every `infeasible` → INFEASIBLE (in the relaxation), var-counts within ±25% of target, reproducible, MPS export OK.
- **Full test suite**: the 7 new variants pass 37/37 each (+262 passing assertions overall).
- Two edge-case sizing bugs found in review were fixed and re-verified (`tracking_error` `Uniform` range inversion at `n_assets==5` / `>300`; `p_median` infeasible mode at the minimum customer count `C==2`).

## Known issue (pre-existing, not introduced here)

`test/runtests.jl` has **6 failures** at line 33 (the ±25% var-count tolerance, fixed seed 0) in untouched variants (`airline_crew`, `cutting_stock`, `network_flow/standard`, `scheduling`, `supply_chain`). Verified **identical on clean `main`** (7c823a9), so they predate this change. The 7 new variants all pass.

## Commit layout

One commit per problem type (category), each bundling the variant file(s) with their wiring so every commit leaves the module loadable; cross-cutting docs in a final commit.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01U76xb1nGaJdYtPzPmej1mC

### CODE REVIEW

**Comment by flixpar** [comment]
@codex review

File: src/problem_types/assignment/workload_balance.jl [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+
+    # --- Dimension sizing: total = W*T + 1, with T > W ---
+    # Pick a worker:task ratio r = T/W in (1.5, 3.0). Then W*T = r*W^2 ≈ target-1
+    # => W ≈ sqrt((target-1)/r), T ≈ r*W.
+    eff = max(4, target_variables - 1)
+    ratio = 1.5 + 1.5 * rand()  # tasks are 1.5x .. 3.0x the worker count
+    n_workers = max(2, round(Int, sqrt(eff / ratio)))
+    n_tasks = max(n_workers + 1, round(Int, eff / n_workers))
+    # One fine-tune pass so W*T tracks target after the integer rounding above.
+    n_tasks = max(n_workers + 1, round(Int, eff / n_workers))
```
[the remark this answers is not in the document]
Comment (flixpar) [reply src/problem_types/assignment/workload_balance.jl]: Fixed in 02dbba8. Removed the duplicate `n_tasks` line (and its "fine-tune pass" comment) — `n_workers` and `eff` are unchanged between the two assignments, so the second was a no-op recompute of the same value. Every generated instance is byte-identical (`workload_balance.jl:91`).

File: src/problem_types/assignment/workload_balance.jl [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+        # Distribute total_cap across workers with mild heterogeneity, then rescale
+        # exactly so the sum equals total_cap (keeps the strict inequality intact).
+        raw = [0.7 + 0.6 * rand() for _ in 1:n_workers]
+        s = sum(raw)
+        for w in 1:n_workers
+            capacities[w] = total_cap * raw[w] / s
+        end
+        # Guard: ensure strict shortfall even after rounding noise.
+        scale = (frac * total_load) / sum(capacities)
+        capacities .*= scale
```
[the remark this answers is not in the document]
Comment (flixpar) [reply src/problem_types/assignment/workload_balance.jl]: Fixed in 02dbba8. Removed the rescale guard — `sum(capacities) == total_cap == frac*total_load` by construction, so `scale` was always 1.0, and the trailing `floor(...)` already guarantees the strict pigeonhole shortfall. Added a short comment in its place explaining why no rescale is needed (`workload_balance.jl:159-161`).

File: src/problem_types/assignment/workload_balance.jl [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+
+            # Greedy LPT: assign heaviest tasks first to the least-loaded eligible
+            # worker. Compute resulting per-worker loads; if any exceeds its cap,
+            # raise that cap so the concrete integer assignment is admissible.
+            order = sortperm(loads; rev=true)
+            assigned_load = zeros(Float64, n_workers)
+            for j in order
+                cands = [w for w in 1:n_workers if eligible[w, j]]
+                # (Guaranteed non-empty by the coverage fix above.)
+                best = cands[argmin([assigned_load[w] for w in cands])]
```
[the remark this answers is not in the document]
Comment (flixpar) [reply src/problem_types/assignment/workload_balance.jl]: Fixed in 02dbba8. Switched to the two-argument `argmin(w -> assigned_load[w], cands)` — equivalent (both return the first minimizer) and avoids the temporary array (`workload_balance.jl:188`).

File: src/problem_types/network_flow/generalized_flow.jl [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+    # Generalized (multiplicative) conservation at intermediate nodes:
+    #   sum_in g*f  ==  sum_out f
+    for v in 1:prob.n_nodes
+        v == src && continue
+        v == snk && continue
+        in_arcs = [arc for arc in arcs if arc[2] == v]
+        out_arcs = [arc for arc in arcs if arc[1] == v]
+        (isempty(in_arcs) && isempty(out_arcs)) && continue
+        inflow = isempty(in_arcs) ? AffExpr(0.0) : sum(prob.gains[arc] * f[arc] for arc in in_arcs)
+        outflow = isempty(out_arcs) ? AffExpr(0.0) : sum(f[arc] for arc in out_arcs)
```
[the remark this answers is not in the document]
Comment (flixpar) [reply src/problem_types/network_flow/generalized_flow.jl]: Fixed in 02dbba8. Both empty-arc fallbacks now use `0.0`. A node with both in- and out-arc sets empty is skipped one line above, so at least one side of the constraint is always an affine expression with variables, and JuMP handles `0.0 == affexpr` fine (`generalized_flow.jl:305-306`).

File: src/problem_types/vehicle_routing/cvrp.jl [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+        cost_per_km = rand(1.5:0.2:3.5)
+        n_clusters = rand(5:8)
+    end
+
+    # --- Depot near grid center ---
+    depot_location = (grid_size * (0.4 + 0.2 * rand()), grid_size * (0.4 + 0.2 * rand()))
+
+    # --- Customers clustered into a few neighborhoods ---
+    cluster_centers = [(grid_size * rand(), grid_size * rand()) for _ in 1:n_clusters]
+    cluster_spread = min(grid_size, grid_size) / (2.5 * n_clusters)
```
[the remark this answers is not in the document]
Comment (flixpar) [reply src/problem_types/vehicle_routing/cvrp.jl]: Fixed in 02dbba8. Simplified to `cluster_spread = grid_size / (2.5 * n_clusters)` (`cvrp.jl:129`).

File: src/problem_types/vehicle_routing/cvrp.jl [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+            dist[i, j] = round(d * cost_per_km * (0.95 + 0.1 * rand()), digits=2)
+        end
+    end
+
+    # --- Resolve feasibility intent ---
+    if feasibility_status == feasible
+        # Guarantee K*Q >= total_demand with margin, K <= N, max demand <= Q.
+        # K is already <= N and sized from demand; widen Q if the margin is thin.
+        if n_vehicles * vehicle_capacity < total_demand * 1.15
+            vehicle_capacity = round(total_demand * 1.15 / n_vehicles, digits=2)
```
[the remark this answers is not in the document]
Comment (flixpar) [reply src/problem_types/vehicle_routing/cvrp.jl:200]: Fixed in 02dbba8. Good catch — you're right that aggregate `K*Q >= total_demand` is necessary but not sufficient for the *integer* CVRP, since the demands must also partition into K routes of capacity Q. The `feasible` branch now certifies this with a first-fit-decreasing bin-packing check (new `_ffd_bin_count`), raising `Q` until the demands pack into `<= K` bins; with `N >= K` this extends to exactly K non-empty routes, so a concrete integer routing provably exists (the same integer-solution-construction approach `assignment/workload_balance` already uses). Raising Q only loosens the relaxation, so LP feasibility is preserved too, and the docstring is updated (no longer scoped to the LP relaxation only).

Verification: 100/100 `feasible` CVRP MIPs (`relax_integer=false`, targets {60,100,200,400} x seeds 1-25) now solve to OPTIMAL under HiGHS.

Worth noting for context: across 3600 scanned feasible instances the existing log-normal demand sizing already kept every instance bin-packable into <= K bins (0 violations), so in practice Q is never actually raised — the guard is purely defensive against the worst case (FFD's 1.222x bound vs. the generator's 1.15x slack when `K < N`). It consumes no RNG and is a strict no-op on all sampled instances, so the generated corpus and reproducibility are unchanged (`cvrp.jl:184-193`).

**Comment by flixpar** [comment]
## Review feedback addressed (02dbba8)

All 6 automated review comments are fixed:

**Cleanups (gemini-code-assist):**
- `cvrp`: `min(grid_size, grid_size)` → `grid_size`
- `generalized_flow`: `AffExpr(0.0)` → `0.0` for empty-arc fallbacks
- `workload_balance`: dropped a duplicate no-op `n_tasks` recompute, dropped a redundant capacity rescale (`scale` was always `1.0`), and switched to idiomatic two-arg `argmin`

**Correctness (chatgpt-codex P2):**
- `cvrp` `feasible` now certifies demands partition into `<= K` routes via a first-fit-decreasing guard, so the integer MIP is feasible — not just the LP relaxation. Purely defensive: a strict no-op across 3600 scanned instances, consumes no RNG, so the corpus and all seeds are byte-identical.

**Verification:** 100/100 feasible CVRP MIPs (`relax_integer=false`) solve to OPTIMAL under HiGHS. Full suite at 2217 passing; the 6 remaining failures are a pre-existing variable-count-tolerance flake in untouched variants (confirmed identical on the clean base).
